### ISSUE TITLE
Question about tags, disaster recovery, and rollbacks

### ISSUE THREAD

**Issue by jantman** [dlandon/zoneminder#134]
@dlandon this is an amazing chunk of work, thank you so much! I first started using ZM a few years ago and have it installed using native OS packages now because I couldn't find a strongly-recommended Docker image at the time, but this has clearly changed, and I'm looking to migrate my current setup to Docker.

I'm a bit curious/confused, though, in seeing that this repo doesn't have any tags/releases, and that the image on the Docker Hub only has the ``latest`` tag.

Given that, do you have a recommendation about how to handle disaster recovery (i.e. I have a backup of my ZM data and a new machine/hard drive, I want to exactly reproduce my previous setup on the new machine) or how to handle rolling back from an upgrade that I didn't like, or didn't work?

**Comment by dlandon** [comment]
The docker itself does not have a need for versioning.  Zoneminder is updated from the ppa to get the latest version of 1.34.x when the docker is first started.  The docker is not built with a particular 1.34.x version.  This is done so the docker does not have to be re-built on every update of the ZM 1.34 version.  ZM has a built in script that is run to make db updates appropriate for the current version.

As far as disaster recovery, I'm not sure what you are looking for as an answer.  The db and events are kept outside the docker so the docker can be removed and re-installed and all settings and events will be kept.  If the db or events get damaged, you will need to recover a backup.

The ability to roll back an upgrade is not something I will spend any time implementing (not worth the effort) and I'm not even sure ZM supports rolling back the db.

**Comment by jantman** [comment]
Thanks, @dlandon, I guess that answers my questions. I'm likely going to have to fork this and make some modifications, as I really want to separate recovering my system and upgrading (and also not have automatic upgrades). Essentially, my current practice (bare install on Debian OS) is to pin all of my package versions, and back up my database and configs every night. The machine is managed via config management (Puppet), so if it dies, I can rebuild to the exact same versions I had before and restore everything, and separate the upgrade process from disaster recovery.

Regarding rollbacks, I wouldn't expect that to be implemented in the docker image. All I really want is the ability to back everything up, try upgrading ZM, and if something doesn't work or I happen to hit a bug in the new version, I delete the configs and DB, restore from backup, and reinstall the older version (i.e. bring everything back to exactly how it was before the upgrade, losing any new data).

Thanks,
Jason

**Comment by alexyao2015** [comment]
@jantman 
Very much agree with this. It's best practice to separate build and installation from runtime with containers. This image seems to break fundamental Docker rules which makes reproduction and usage difficult to manage. In the current state, the image itself is being modified during runtime which really isn't ideal for reproducible containers. Docker is not a VM and should not be treated as one. 

I'm also looking to build a new container from scratch once I'm able to understand how to properly build and install zoneminder. In the current form, I just do not feel comfortable using it, knowing that the image itself is mutable.

**Comment by jantman** [comment]
@alexyao2015 FYI, I'm still using 1.32 but planning on upgrading in the next few weeks. I ended up building an image based off on QuantumObject/docker-zoneminder, because that image comes much closer to following Docker's best practices, and works a lot more like how people expect Docker images to work. I did end up forking it ( https://github.com/jantman/docker-zoneminder ) because the original image's scripts require root MySQL access, which shouldn't be necessary if the database already exists.
